evcxr/src/module.rs

```diff
@@ -266,27 +272,76 @@ overflow-checks = true
     }
 }
 
+fn tee(mut stream: impl Read, callback: impl Fn(&str)) -> Result<String, Error> {
+    let mut buf_stream = BufReader::new(&mut stream);
+    let mut out: String = String::new();
+    let mut buff = String::new();
+    // I dont like infinite loops
+    loop {
+        let bytes_read = buf_stream.read_line(&mut buff)?;
+        if bytes_read == 0 {
+            return Ok(out);
+        }
+        //TODO:Is state variable needed for the callback, incase the callback needs to buffer some info?
+        callback(&buff);
+        out.push_str(&buff);
+        buff.clear();
+    }
+}
+
+fn get_callback(code_block: &CodeBlock) -> impl Fn(&str) {
+    //Idea is to add some check on the code_block to decide if additional logging is necessary
```

Comment:
Sorting extra configuration information on the CodeBlock doesn't seem like the right place to me. You mention below that the function signature looks complex if you change run_cargo to take a callback, but is it actually that bad?

Reply:
`fn subfunc<F: 'static + FnMut() -> () + Send>(f: F) {`
details of it are discussed [here](https://users.rust-lang.org/t/passing-closure-received-as-argument-to-thread/55567/3).

I don't really mind adding this. But we need to create the callback(s) and pass it (we might need one callback for `stderr` and one for `stderr`) somewhere. And shouldn't we be adding the callback only when the codeblock that is executed is a  `:dep`. So how can we decide the callback without a reading the code_block.  
Without this check I see lot of logs that is unnecessary logs printed.

Also I am not clear about how the jupyter flow really works. I was not able to connect flow of control to the `run_cargo` fn that I am editing here.

Reply:
I think `run_cargo` should only need to accept callback for stderr. Processing stdout can continue to be handled entirely within `run_cargo` as it is now.

I think the callback would need to be passed through the following functions:
* `run_cargo`
* `Module::compile`

`Module::compile` is called from `EvalContext::try_run_statements`. One of the arguments to `try_run_statements` is `callbacks: &mut EvalCallbacks`. `EvalCallbacks` seems like a perfect place to add the callback.

The REPL currently calls `context.execute`, but could be changed to call `context.execute_with_callbacks`, which is the variant that takes an `EvalCallbacks` argument.

Regarding whether to only print stderr when a `:dep` command has been issued, I'm not sure. It might be simpler to just always print progress from the compiler. Or perhaps decide whether to print based on how long things are taking - e.g. for the first second, don't print anything, but then if we haven't finished compiling within a second print progress. But probably see how it looks first when doing it unconditionally. It might depend how chatty cargo is when doing normal compilation. You could also potentially filter out lines that you don't want to avoid them being printed.

Reply:
> I think `run_cargo` should only need to accept callback for stderr. Processing stdout can continue to be handled entirely within `run_cargo` as it is now.
> 
This makes sense. 👍

> I think the callback would need to be passed through the following functions:
> 
> * `run_cargo`
> * `Module::compile`
> 
> `Module::compile` is called from `EvalContext::try_run_statements`. One of the arguments to `try_run_statements` is `callbacks: &mut EvalCallbacks`. `EvalCallbacks` seems like a perfect place to add the callback.
> 
> The REPL currently calls `context.execute`, but could be changed to call `context.execute_with_callbacks`, which is the variant that takes an `EvalCallbacks` argument.
> 

I tried passing `EvalCallbacks` from newly created `Repl::execute_with_callbacks`  as you suggested.
But while passing the callback to the thread the callback I get an error saying `cannot be shared between threads safely`
![error](https://user-images.githubusercontent.com/8631902/220268358-c1db2c6d-6eb7-40e0-8b90-92122354677f.png)

I am trying to see how I can get this error fixed.

Reply:
If you wrap the callback in `Arc<Mutex<..>>` you should be able to share it between threads. For example:

```rust
use std::sync::Arc;
use std::sync::Mutex;

struct Foo {
    callback: Arc<Mutex<dyn FnMut() -> i32 + Send>>,
}

fn main() {
    let f = Foo {
        callback: Arc::new(Mutex::new(|| 42)),
    };
    let t = std::thread::spawn(move || {
        let callback = &mut *f.callback.lock().unwrap();
        let v = (callback)();
        println!("{v}");
    });
    t.join().unwrap();
}
```

It's fine to pass `EvalCallbacks` into `run_cargo`, but I'd probably suggest avoiding passing it to the separate thread, because then you'd need to wrap the whole of `EvalCallbacks` in `Arc<Mutex<...>>` rather than just the callback.

If your callback ends up not needing any state, then it can probably be a `Fn` rather than a `FnMut`. If that's the case, then you probably don't need the `Mutex`, just the `Arc`. e.g.:

```rust
use std::sync::Arc;

struct Foo {
    callback: Arc<dyn Fn() -> i32 + Send + Sync>,
}

fn main() {
    let f = Foo {
        callback: Arc::new(|| 42),
    };
    let t = std::thread::spawn(move || {
        let callback = &*f.callback;
        let v = (callback)();
        println!("{v}");
    });
    t.join().unwrap();
}
```